### PR TITLE
Change how runmake v2 tests are executed

### LINKED ISSUE #125726: run-make-support: rename `tmp_dir()` and clarify its semantics
Most command wrappers assume output artifacts will be emitted to `tmp_dir()` (which calculates `TMPDIR` as is used in tools.mk), which is a scratchpad directory that is intended to be unique for each run-make test so they don't interfere with each other, and that output artifact emission is contained to the directory. Typically command wrappers will have `--out-dir=tmp_dir()` presets.

The naming of `tmp_dir()` however is pretty bad -- it being a "temporary directory" is not to be confused with `env::temp_dir`.

Do note, however, that tools.mk setting `TMPDIR` has other effects:

> On Unix, returns the value of the TMPDIR environment variable if it is set

I don't know if this is intentional, but if it is intended to even control where e.g. codegen artifacts are emitted, then setting `TMPDIR` does nothing on Windows: we also need to set `TMP`/`TEMP` on Windows.

### LINKED ISSUE #126080: run-make: do not run test with the test sources directory as its CWD
The original infra PR setting up `rmake.rs` version of run-make tests copied over the logic for the legacy `Makefile` version, which ran the test with the *test sources* directory as the CWD, which is *Very Naughty* because if you aren't careful with relative paths, you suddenly can produce output artifacts into your *test sources* directory. Oops.

We should stop doing that.

cc @Kobzol as they wanted to work on a fix for this.

### PR DESCRIPTION
**Pull request by Kobzol**
This PR makes execution of v2 runmake tests more sane, by executing each test in a temporary directory by default, rather than running it inside `tests/run-make`. This will have.. a lot of conflicts.

Fixes: https://github.com/rust-lang/rust/issues/126080
Closes https://github.com/rust-lang/rust/issues/125726, because it removes `tmp_dir`, lol.

r? @jieyouxu

try-job: x86_64-msvc

### CODE REVIEW

**Comment by Kobzol** [comment]
This is now ready for review (best reviewed commit-by-commit, of course).

@rustbot ready

File: src/tools/compiletest/src/runtest.rs [hunk 1]
```diff
@@ -3458,6 +3458,21 @@ impl<'test> TestCx<'test> {
         let rmake_out_dir = base_dir.join("rmake_out");
         create_dir_all(&rmake_out_dir).unwrap();
 
+        // Copy all input files (apart from rmake.rs) to the temporary directory,
+        // so that the input directory structure from `tests/run-make/<test>` is mirrored
+        // to the `rmake_out` directory.
+        for path in walkdir::WalkDir::new(&self.testpaths.file).min_depth(1) {
+            let path = path.unwrap().path().to_path_buf();
+            if path.file_name().map(|s| s != OsStr::new("rmake.rs")).unwrap_or(false) {
```
Comment (jieyouxu) [inline src/tools/compiletest/src/runtest.rs:3466]: Suggestions:

- `s != "rmake.rs"`, I think there's an impl for `PartialEq` between `&str` and `OsStr`?
- `file_name().map().unwrap_or(false)` can be written as `file_name().is_some_and(|n| n != "rmake.rs")`

File: src/tools/run-make-support/src/lib.rs [hunk 2]
```diff
@@ -315,6 +315,28 @@ pub fn assert_not_contains(haystack: &str, needle: &str) {
...
+///
+/// What this function does:
+/// 1) Creates a temporary directory (`tmpdir`)
+/// 2) Copies all files from the current directory to `tmpdir`
+/// 3) Changes the current working directory to `tmpdir`
+/// 4) Calls `callback`
+/// 5) Switches working directory back to the original one
+/// 6) Removes `tmpdir`
+/// Switch current working directory to a temporary directory
+pub fn run_in_tmpdir<F: FnOnce()>(callback: F) {
```
Comment (jieyouxu) [inline src/tools/run-make-support/src/lib.rs:328]: I like this helper.

File: tests/run-make/c-link-to-rust-dylib/rmake.rs [hunk 3]
```diff
@@ -5,29 +5,23 @@
...
-            .library_search_path(tmp_dir())
-            .run();
+        cc().input("bar.c").arg("-lfoo").output("bar").library_search_path(cwd()).run();
     }
 
     run("bar");
 
     let expected_extension = dynamic_lib_extension();
-    read_dir(tmp_dir(), |path| {
+    read_dir(std::env::current_dir().unwrap(), |path| {
```
Comment (jieyouxu) [inline tests/run-make/c-link-to-rust-dylib/rmake.rs:24]: Nit: should this use the `cwd()` helper you introduced in the support lib?

File: tests/run-make/const-prop-lint/rmake.rs [hunk 4]
```diff
@@ -2,12 +2,12 @@
...
 use std::fs;
 
-use run_make_support::{rustc, tmp_dir};
+use run_make_support::rustc;
 
 fn main() {
     rustc().input("input.rs").run_fail_assert_exit_code(1);
 
-    for entry in fs::read_dir(tmp_dir()).unwrap() {
+    for entry in fs::read_dir(std::env::current_dir().unwrap()).unwrap() {
```
Comment (jieyouxu) [inline tests/run-make/const-prop-lint/rmake.rs:10]: Nit: also `cwd()`?

File: tests/run-make/CURRENT_RUSTC_VERSION/rmake.rs [hunk 5]
```diff
@@ -10,11 +10,11 @@ use run_make_support::{aux_build, rustc};
...
-    let mut stable_path = PathBuf::from(env!("TMPDIR"));
-    stable_path.push("libstable.rmeta");
-
-    let output =
-        rustc().input("main.rs").emit("metadata").extern_("stable", &stable_path).command_output();
+    let output = rustc()
+        .input("main.rs")
+        .emit("metadata")
+        .extern_("stable", "libstable.rmeta")
+        .command_output();
```
Comment (jieyouxu) [inline tests/run-make/CURRENT_RUSTC_VERSION/rmake.rs:17]: Reminder (for myself): pre-existing; this needs to use `run()` to check exit status.

File: tests/run-make/notify-all-emit-artifacts/rmake.rs [hunk 6]
```diff
@@ -19,7 +17,7 @@ fn main() {
             .codegen_units(1)
             .json("artifacts")
             .error_format("json")
-            .incremental(&inc_dir)
+            .incremental(&std::env::current_dir().unwrap())
```
Comment (jieyouxu) [inline tests/run-make/notify-all-emit-artifacts/rmake.rs:20]: Nit: `cwd()`?

File: tests/run-make/notify-all-emit-artifacts/rmake.rs [hunk 7]
```diff
@@ -35,7 +33,7 @@ fn main() {
             .codegen_units(2)
             .json("artifacts")
             .error_format("json")
-            .incremental(&inc_dir)
+            .incremental(&std::env::current_dir().unwrap())
```
Comment (jieyouxu) [inline tests/run-make/notify-all-emit-artifacts/rmake.rs:36]: Nit: `cwd()`

File: tests/run-make/rustdoc-scrape-examples-macros/rmake.rs [hunk 8]
```diff
@@ -41,8 +40,8 @@ fn main() {
         .crate_name("ex")
         .crate_type("bin")
         .output(&out_dir)
-        .extern_(crate_name, tmp_dir.join(format!("lib{crate_name}.rlib")))
-        .extern_(proc_crate_name, tmp_dir.join(dylib_name.trim()))
+        .extern_(crate_name, format!("lib{crate_name}.rlib"))
```
Comment (jieyouxu) [inline tests/run-make/rustdoc-scrape-examples-macros/rmake.rs:43]: Reminder (for myself): this should use `rust_lib_name()`.

**Review by jieyouxu — APPROVED** [review]
Love the changes, takes away a lot of visual noise in the test.

**Comment by jieyouxu** [comment]
Changes look good, only have minor non-blocking nits, feel free to r=me with or without addressing them.

File: src/tools/compiletest/src/runtest.rs (hunk 1 above)
Comment (Kobzol) [reply src/tools/compiletest/src/runtest.rs:3466]: TIL about that `PartialEq` impl, thanks!

**Comment by Kobzol** [comment]
Fixed the nits (apart from the `command_output`, because that requires a bit more work across `run-make-support`, we talked about this before - adding some custom `Output` struct for easy stdout/stderr asserting etc.). I also included one more comment based on the rust lib path cleanup, as we no longer need two functions for name/path of static/dynamic/rlib libraries.

Let me know if the last two commits are OK for you.

@bors rollup=never

**Review by jieyouxu — APPROVED** [review]

**Comment by jieyouxu** [comment]
Looks good, r=me after CI is green

**Comment by Kobzol** [comment]
Missed a few imports in non-Linux tests.

@bors r=jieyouxu

**Comment by Nilstrieb** [comment]
@bors r-
bors what the hell are you doing are you are testing https://github.com/rust-lang/rust/pull/125966#issuecomment-2156034298 already

**Comment by Nilstrieb** [comment]
oh this is supposed to be a try build... the queue disagrees :)

**Comment by BoxyUwU** [comment]
pietro commanded me to:
@bors retry

**Comment by Nilstrieb** [comment]
it's not in the queue, so
@rustbot author

**Pull request merged**